Tasks/DownloadBuildArtifacts/main.ts

```diff
@@ -0,0 +1,151 @@
+var path = require('path')
+var url = require('url')
+
+import * as tl from 'vsts-task-lib/task';
+import { IBuildApi } from 'vso-node-api/BuildApi';
+import { IRequestHandler } from 'vso-node-api/interfaces/common/VsoBaseInterfaces';
+import { WebApi, getHandlerFromToken } from 'vso-node-api/WebApi';
+
+import * as models from 'item-level-downloader/Models';
+import * as engine from 'item-level-downloader/Engine';
+import * as providers from 'item-level-downloader/Providers';
+import * as webHandlers from 'item-level-downloader/Providers/Handlers';
+
+tl.setResourcePath(path.join(__dirname, 'task.json'));
+
+async function main(): Promise<void> {
+    var promise = new Promise<void>(async (resolve, reject) => {
+        var buildType: string = tl.getInput("buildType", true);
+        var isCurrentBuild: boolean = buildType.toLowerCase() === 'current';
+        var projectId: string = isCurrentBuild ? tl.getVariable("System.TeamProjectId") : tl.getInput("project", true);
+        var definitionId: string = isCurrentBuild ? '' : tl.getInput("definition", true);
+        var buildId: number = parseInt(isCurrentBuild ? tl.getVariable("Build.BuildId") : tl.getInput("buildId", true));
+        var downloadPath: string = tl.getInput("downloadPath", true);
+        var downloadType: string = tl.getInput("downloadType", true);
+
+        var endpointUrl: string = tl.getVariable("System.TeamFoundationCollectionUri");
+        var accessToken: string = tl.getEndpointAuthorizationParameter('SYSTEMVSSCONNECTION', 'AccessToken', false);
+        var credentialHandler: IRequestHandler = getHandlerFromToken(accessToken);
+        var webApi: WebApi = new WebApi(endpointUrl, credentialHandler);
+        var debugMode: string = tl.getVariable('System.Debug');
+        var isVerbose: boolean = debugMode ? debugMode.toLowerCase() != 'false' : false;
+        var parallelLimit: number = +tl.getVariable("release.artifact.download.parallellimit");
```

Comment:
This looks like a release-specific variable to me.

Reply:
@scottdallamura This is the variable we support for Release as part of current download artifact and has been added for backward compact, if Build has a similar variable we can pick that up in the task based on host type.

Reply:
Why was it implemented as a variable?  Shouldn't it be an input on the task?

The existing DownloadBuildArtifact task does not use this variable, as far as I can tell - at least, not in releases/m122.  Can you explain the compat scenario here?

This doesn't seem like something we want to carry forward.

Reply:
@scottdallamura - we were referring to current download artifact implementation we have within Release extension in agent:
https://github.com/Microsoft/vsts-agent/blob/28790c1ab9ffc9fa597fc1477401227d21211209/src/Agent.Worker/Release/Artifacts/BuildArtifact.cs#L235

we will be moving the download artifact logic into the task. we want to carry forward the functionality we offered in the existing download artifact logic into the task. 

we can choose to expose it as a task input if required as another minor version patch.

Reply:
Are you going to be migrating release definitions to use this task?  If not, then I don't see a compat issue, and no reason to pollute the new task with old, weird behavior.  It should require no more work to read from the input than to read from the variable - it's literally tl.getInput instead of tl.getVariable.

Reply:
No migration of existing RDs is required. The task will not be added to RD, it will be added by RM into the job automatically.

We can still go ahead and make it a task input & RM can populate the input based on variable set. Agree it is minimal work.

But at this point, we don't want to expose too many task inputs. We can always choose to expose it later if required with a minor version patch.

Reply:
As long as the task references a release-specific variable, it's not production-ready.